server/jobs/github_search.py

```diff
@@ -0,0 +1,196 @@
+    r.raise_for_status()
+    return r.json()
+
+def search_request(search_line, language, logger, access_token, page=1):
+    """ Query https://api.github.com/search/code """
+    search_query = ('"{line}" in:file language:{lang}'
+                    .format(line=search_line, lang=language))
+    if page > 6:
+        logger.info("Stopping at page 6")
+        return []
+    try:
+        response = requests.get(
+            url="https://api.github.com/search/code",
+            params={
+                "q": search_query,
+                "page": "{}".format(page),
+                "per_page": "50",
+                "sort": "updated"
+            },
+            headers={
+                "Authorization": "Bearer {}".format(access_token),
+                # "Accept": "application/vnd.github.v3.text-match+json"
+                # if partial text matches data is useful
+            },
+        )
+        data = response.json()
+        response.raise_for_status()
+        remaining = response.headers.get('X-RateLimit-Remaining')
+        reset = response.headers.get('X-RateLimit-Reset')
+        if not remaining and reset:
+            seconds_wait = reset - time.gmtime()
+            friendly_limit = arrow.utcnow().replace(seconds=seconds_wait * 5).humanize()
+            logger.warning(("Rate limited on page {}. Quota refreshing {}"
+                            .format(page, friendly_limit)))
+            return []
+        elif 'items' not in data:
+            logger.warning("Could not query page {}: {}".format(page, data))
+            return []
+        return data
+    except requests.exceptions.RequestException as e:
+        if response.status_code == 403:
+            logger.warning("No Token/Rate Limit: {}".format(response.text))
+        logger.warning("HTTP Error when fetching {}: \n {}".format(search_query, e))
+        return
+
+def get_all_results(search_line, language, logger, access_token):
+    logger.info("Performing search for {}".format(search_line))
+    first_req = search_request(search_line, language, logger, access_token)
+    if not first_req:
+        logger.warning("Failure to perform first search")
+        return
+    total_count = first_req.get('total_count')
+    if not total_count:
+        logger.warning("There were no matching results on GitHub")
+        return
+    logger.info("Repos Found {}".format(total_count))
+    pages = range(2, min(math.ceil(total_count/50)+1, 5))
+    data_set = first_req['items']
+    for page_num in pages:
+        logger.info("Fetching page {}".format(page_num))
+        time.sleep(8)
+        results = search_request(search_line, language, logger, access_token,
+                                 page=page_num)
+        if not results or 'items' not in results:
+            logging.warning("Request {} failed {}".format(page_num, results))
+            return data_set
+        else:
+            data_set.extend(results['items'])
+    return data_set
+
+def download_repos(repos):
```

Comment:
I know this is for future use, but this should probably take a directory path and the caller can make a tmpdir if they want (and they might not want to)